### DISCUSSION TITLE
Storing a setting from within phoebus?

### DISCUSSION THREAD

**Discussion by daykin** [ControlSystemStudio/phoebus#3212] [Q&A]
What is the preferred way to store a setting in settings.ini from phoebus? When the user opts in or out of my analytics plugin, I want to store the value.

**Comment by georgweiss** [comment]
Sorry for the delay.
An ```AppInstance``` may override ```save(Memento memento)``` to store something when application is closed, and pick up saved value in ```restore(Memento memento)```.

To save/restore outside the restore/save lifecycle you may need to roll your own, preferably in a directory used by the lifecycle. ```Locations.user()``` returns the (hidden) folder containing data created through the lifecycle API. Defaults to ```~/.phoebus```.  

As for settings.ini... I would avoid manipulating it from Phoebus.

**Comment by kasemir** [comment]
Yes, we don't save preferences other than the "memento".
So you can store the state of a panel in the UI, but once you close that panel, nothing will be saved for that now non-existing panel.

We used to store the preferences like this:
1) Read preferences defaults
2) Check for persisted settings in `$HOME/.phoebus/...` and read those
2) Read preferences from *.ini files
3) Write end result via `FileSystemPreferences` to `$HOME/.phoebus/...`

Whenever you then changed the settings via `java.util.prefs`, the updated settings would again be written.

That resulted in two issues:
a) If you once ran with `-settings /some/settings.ini`, the result got persisted.
  If you next ran without `-settings ...`, you would still get the settings from that last used `settings.ini`, somewhat unexpected.
  This could really bite you if ran `alarm-server -settings /file/for/alarms.ini` and then the GUI: The GUI would now use settings that were meant for the alarm server!!
b) The `FileSystemPreferences` code which we had basically copied from the Linux preference implementation to get the exact same behavior on Linux, Windows and Mac contained calls to deprecated API.


So in #3041 we simplified. Remove our copy of `FileSystemPreferences` and simply do this:
1) Read preferences defaults
2) Read preferences from *.ini files
..done

You thus get the built-in defaults or what's added via a `settings.ini` in your product and finally `-settings ...` command line switches. Nothing else is unexpectedly read from some other files. But there's also nothing persisted any longer.
